dotnet/src/Planners/Planners.Core/Handlebars/Models/HandlebarsParameterTypeMetadata.cs

```diff
@@ -0,0 +1,115 @@
+    public bool IsComplex { get; set; } = false;
+
+    /// <summary>
+    /// If this is a complex type, this will contain the properties of the complex type.
+    /// </summary>
+    [JsonPropertyName("properties")]
+    public List<SKParameterMetadata> Properties { get; set; } = new();
+
+    // Override the Equals method to compare the property values
+    public override bool Equals(object obj)
+    {
+        // Check if the obj is null or of a different type
+        if (obj == null || obj.GetType() != this.GetType())
+        {
+            return false;
+        }
+
+        // Cast the obj to HandlebarsParameterTypeMetadata
+        var other = (HandlebarsParameterTypeMetadata)obj;
+
+        // Compare the Name and IsComplex properties
+        if (this.Name != other.Name || this.IsComplex != other.IsComplex)
+        {
+            return false;
+        }
+
+        // Compare the Properties lists using a helper method
+        return ArePropertiesEqual(this.Properties, other.Properties);
+    }
+
+    // A helper method to compare two lists of SKParameterMetadata
+    private static bool ArePropertiesEqual(List<SKParameterMetadata> list1, List<SKParameterMetadata> list2)
+    {
+        // Check if the lists are null or have different lengths
+        if (list1 == null || list2 == null || list1.Count != list2.Count)
+        {
+            return false;
+        }
+
+        // Compare the elements of the lists by comparing the Name and ParameterType properties
+        for (int i = 0; i < list1.Count; i++)
+        {
+            if (!list1[i].Name.Equals(list2[i].Name, System.StringComparison.Ordinal) || !list1[i].ParameterType!.Equals(list2[i].ParameterType))
+            {
+                return false;
+            }
+        }
+
+        // If all elements are equal, return true
+        return true;
+    }
+
+    // Override the GetHashCode method to generate a hash code based on the property values
+    public override int GetHashCode()
+    {
+        // Use a prime number to combine the hash codes of the properties
+        int hash = 17;
+
+        // Use the default string hash code for the Name property
+        hash = (hash * 31) + (this.Name?.GetHashCode() ?? 0);
+
+        // Use the default bool hash code for the IsComplex property
+        hash = (hash * 31) + this.IsComplex.GetHashCode();
+
+        // Use a helper method to compute the hash code of the Properties list
+        hash = (hash * 31) + GetPropertiesHashCode(this.Properties);
+
+        return hash;
+    }
+
+    // A helper method to compute the hash code of a list of SKParameterMetadata
+    private static int GetPropertiesHashCode(List<SKParameterMetadata> list)
+    {
+        // Use a prime number to combine the hash codes of the elements
+        int hash = 19;
+
+        // Use the default SKParameterMetadata hash code for each element
+        foreach (var item in list)
+        {
+            hash = (hash * 37) + GetPropertyHashCode(item);
```

Comment:
No need in this method if SKParameterMetadata implements GetHashCode method. @stephentoub Is there an easier way in .NET to perform hash calculations?

Reply:
System.HashCode is the recommended way to combine multiple hash codes into a single one, e.g.
```C#
public override int GetHashCode()
{
    HashCode hc = default;
    hc.Add(this.Name);
    hc.Add(this.IsComplex);
    foreach (var item in this.Properties)
    {
        hc.Add(item?.Name);
        hc.Add(item?.ParameterType);
    }
    return hc.ToHashCode();
}
```
That said, also keep in mind that while equality does need to factor in all relevant data, hash code doesn't have to. It just needs to factor in enough to provide a reasonably good-enough distribution.  You could, for example, just have the implementation be:
```C#
public override int GetHashCode() => this.Name?.GetHashCode() ?? 0;
```
and depending on the circumstances, that might be "good enough".

Reply:
Refactored to use HashCode